src/merkle/tiered_smt/storage.rs

```diff
@@ -0,0 +1,69 @@
+use super::{BTreeMap, Content, ContentType, RpoDigest, TreeIndex, Vec};
+use core::{cell::RefCell, fmt};
+
+#[derive(Debug, Default)]
+pub struct Storage {
+    payload: RefCell<BTreeMap<RpoDigest, Vec<u8>>>,
+    nodes: RefCell<BTreeMap<TreeIndex, Content>>,
+}
+
+impl Storage {
+    pub fn insert_payload(&self, key: RpoDigest, payload: Vec<u8>) -> Result<(), StorageError> {
+        self.payload
+            .try_borrow_mut()
+            .map_err(|_| StorageError::PoisonedInternalWrapper)?
+            .insert(key, payload);
+        Ok(())
+    }
+
+    pub fn insert_node(&self, index: TreeIndex, content: Content) -> Result<(), StorageError> {
+        self.nodes
+            .try_borrow_mut()
+            .map_err(|_| StorageError::PoisonedInternalWrapper)?
+            .insert(index, content);
+        Ok(())
+    }
+
+    pub fn insert_node_batch<I>(&self, args: I) -> Result<(), StorageError>
+    where
+        I: IntoIterator<Item = (TreeIndex, Content)>,
+    {
+        self.nodes
+            .try_borrow_mut()
+            .map_err(|_| StorageError::PoisonedInternalWrapper)?
+            .extend(args);
+        Ok(())
+    }
+
+    pub fn peek_node_type(&self, index: &TreeIndex) -> Result<Option<ContentType>, StorageError> {
```

Comment:
The idea is to leave the storage as dumb as possible. It will basically store anything the tree asks, without applying any logic or assumption. There is an equivalent method for the tree that will unwrap to empty, as it is part of the tree logic.